db/db_impl/db_impl.cc

```diff
@@ -3946,6 +3946,16 @@ std::unique_ptr<IterType> DBImpl::NewMultiCfIterator(
     return error_iterator_func(
         Status::InvalidArgument("No Column Family was provided"));
   }
+  if (!_read_options.total_order_seek && !_read_options.auto_prefix_mode) {
+    for (auto* column_family : column_families) {
+      auto cfh = static_cast_with_check<ColumnFamilyHandleImpl>(column_family);
+      auto cfd = cfh->cfd();
+      if (cfd->GetLatestCFOptions().prefix_extractor != nullptr) {
```

Comment:
`GetLatestCFOptions()` requires holding db_mutex_ according to its [comment](https://github.com/facebook/rocksdb/blob/13c758f9869c2b086e9c1f7457c15db93349f258/db/column_family.h#L357).

Also we may want to make this cheap since these are the default read options.

Reply:
@cbi42 

Looks like getting `mutable_cf_options` from a super version is thread safe and should be quite cheap, but I'd like to discuss what's going to be the best way to handle this.

Just calling `cfd->GetReferencedSuperVersion(this)` in `NewMultiCfIterator()` doesn't sound right to me since it could potentially reference a different SV than the ones referenced inside `NewIterators()`, where child iterators are created.

So I am thinking of returning `Status::InvalidArgument("Manual Prefix Iteration is not allowed")` inside `NewIterators()` when `total_order_seek = false`, `auto_prefix_mode = false` and `prefix_extractor != nullptr`. 

However, this would change the existing behavior which could be disruptive to some users. Maybe optional parameter with default false, something like `bool disallow_manual_prefix_iteration = false` can be added to DB::NewIterators() API. I may have to implement this in several places if we go this route, but shouldn't be too bad to add, I think.

Please let me know if you can think of a better way.